List authored posts on team member page

MDZHX opened the pull request:
Added a list of authored blog posts on each team member's page.
A working prototype :)

malsf21 commented:
Looks good so far! I think part of your PR relies on #168 - let me see who's on that right now!

MDZHX commented:
Thanks for the quick reply! I'll also add some preview card prototypes before #156 is implemented.

malsf21 commented:
Changes are looking good Zihan! I think we may wait on #168 depending on when it's going to happen - do you mind touching base with Shounak and seeing what the ETA on that is looking like?

Also, cc: @krashanoff - any thoughts on time complexity / performance implications? This is not my area of expertise (at least, with Ruby haha)

MDZHX commented:
Sure! Will DM Shounak on slack.

MDZHX commented:
I just got a reply from Shounak. He plans to get the basic layout done by this weekend and further polish it afterward. He says that he doesn't have an exact ETA on it right now, though.

krashanoff reviewed (commented):
> any thoughts on time complexity / performance implications? This is not my area of expertise (at least, with Ruby haha)

Definitely isn't my area of expertise either, but at a glance, I don't think this will kill build time on each dev's environment, but it does see a difference in our GitHub builds (the build for this branch was about 30s slower).

![image](https://user-images.githubusercontent.com/7704542/105560198-04486500-5cc8-11eb-86bd-b75e81dbceab.png)

malsf21 commented:
> I just got a reply from Shounak. He plans to get the basic layout done by this weekend and further polish it afterward. He says that he doesn't have an exact ETA on it right now, though.

Gotcha, in the interest of time for now then, let's merge this first, then merge it into Shounak's branch and let him handle using the component. I'll leave a formal review in a moment.

> Definitely isn't my area of expertise either, but at a glance, I don't think this will kill build time on each dev's environment, but it does see a difference in our GitHub builds (the build for this branch was about 30s slower).

Where are you seeing the `30s` difference @krashanoff? I'm peeping the actions build and it's still ~ 1s.

malsf21 commented on _layouts/team-member.html:
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
+        <div class="col">
+            <h2 class="title"> Blog posts that {{ page.firstname }} has written: </h3>
+            <div class="team-links">
+                <ul>
```
If you're using `<ul>`, the children should be `<li>` - you can always restyle them to fit your original vibe, but make sure you match things up semantically!

malsf21 commented on _layouts/team-member.html:
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
+                                    <img  class="post-preview-img" src="{{site.baseurl}}/{{post.image}}" alt="{{post.alt}}" />
+                                </a>
+                                <div class="post-info">
+                                    <h2 class="title">
```
We already have an `<h2>` above; you might want to make this a `<h3>` instead!

malsf21 commented on _layouts/team-member.html:
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
     </div>
+
+    <!-- blog posts -->
+    <div class="row">
+        <div class="col">
```
Is this supposed to be `.row > .col`? There is no row or column haha

MDZHX replied:
Ah, I thought every section of a page was put into a grid and thus has a row/col. Now I know better of the structure :) thanks!

malsf21 commented on _layouts/team-member.html:
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
+                                    <p class="generated-content">{{post.excerpt}}</p>
+                                </div>
+                                </div>
+                            </section>
```
Where is the matching `<section>` tag?

malsf21 reviewed (changes requested):
Okay, left some quick quick comments for you @MDZHX! Nothing too big, just some small semantic things 😊

krashanoff commented:
> Where are you seeing the 30s difference @krashanoff?

In the [Test Jekyll Build](https://github.com/uclaacm/teach-la-website/runs/1747136805) action between the latest commit on this branch and the latest on master. But to reiterate, I saw no detriment on my environment. Further, GitHub actions are a lot slower than Netlify deploys. I'd say this merging this is harmless.

malsf21 commented:
@MDZHX let me know when you have a chance to address the comments, and then I can give you an approving review and we can merge!

MDZHX commented:
@malsf21 Sorry for the late fix! I've been solving a connection problem that prevented me to pull from `master` :( I think it's now ready to merge!

malsf21 commented on _layouts/team-member.html:
```diff
@@ -42,12 +42,47 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
     </div>
+
+    <!-- blog posts -->
+    <h3 class="title"> Blog posts that {{ page.firstname }} has written: </h3>
```
Hm, so I think this is the inverse of the change that we wanted. Keep in mind that headings and subheadings nest; so if you want a heading to be "within" the blog posts that a user has written, the number should increment by one.

In other words - this should be an `<h2>` tag; the heading inside `<div class="post-info">` should be a `h3`!

MDZHX commented:
Thanks for pointing that out! I think now it's fixed.

malsf21 reviewed (changes requested):
Thanks for the quick change!

One last thing I'll mention - can we add a message and/or hide the entire section if the user hasn't written any blog posts? For example, see

https://deploy-preview-213--acm-teach-la-website.netlify.app/team/aghodsian

After this (I've said this a few times, haha) we'll be good to merge!

MDZHX commented:
I looked up some docs and implemented this feature by assigning a boolean flag - not sure if that's a clean way to do it. I also removed the outer div with class="team-links" since I don't see this class in the CSS files. Since I'm not familiar enough with the class structure, I wonder if you think I should remove that?

malsf21 commented:
> I looked up some docs and implemented this feature by assigning a boolean flag - not sure if that's a clean way to do it.

Great question! I think your approach works, but after thinking about it for a hot minute I think there's probably a more elegant solution that involves the [where filter](https://jekyllrb.com/docs/liquid/filters/):

```
{% assign authors_posts = site.posts | where_exp:"post", "SOME_EXPRESSION_YOU_NEED_TO_WRITE" | sort_by:"date" %}
```

This will create an iterable `authors_posts` that only has the authors posts. Then, you can check the size; if it's `> 0`, then render the title + iterate through `authors_posts`. This may also make the process faster!

MDZHX commented:
I'm kind of struggling with the expression to fill in. Each post's `authors` attribute is a list of objects that have first names and last names. Is there a way I can check if an object with a specific first name and last name is in the list?

malsf21 commented:
Yeah, this is a good point - in hindsight, I do wish we designed team pages around a single name (e.g. `Matthew Wang`) instead of a first name / last name. Maybe I'll ping @krashanoff to see if he has any thoughts on this? Worst-case, I'll give this a once over and then just let it merge - may mean that we're due for another refactor on how we handle names.

MDZHX commented:
Thanks! That would be helpful :)

krashanoff commented:
> in hindsight, I do wish we designed team pages around a single name (e.g. Matthew Wang) instead of a first name / last name.

This is what I was originally trying for when I first implemented team pages, but I remember that after some discussion we switched to separating out first and last names. I think it was for sorting members on the team page by name. There may also have been an issue with naming ambiguities.

To help you along with your expression trouble, what if you checked that each first name and last name field matched? For example, assuming that you have access to the author information in your post, or pass it down through some headmatter (accessible via the `page` variable:

```
{% assign authors_posts = site.posts | where_exp:"post", "post.fname == page.fname and post.lname == page.lname" | sort_by:"date" %}
```

> may mean that we're due for another refactor on how we handle names.

We could refactor posts specifically to just reference authors by their team file (e.g., `mwang` corresponds to `_team/mwang.md`). Thoughts?

malsf21 commented:
> We could refactor posts specifically to just reference authors by their team file (e.g., mwang corresponds to _team/mwang.md). Thoughts?

This could be something we tackle later, can you index collections by filename? I've actually ... never tried, haha.

For now, let's limit the scope of this PR. If you're able to get this to work Zihan, that would be appreciated, but if not I'll give your live demo one more test and give it the go-ahead to merge!

MDZHX commented:
I tried Leo's quick fix but that doesn't seem to work based on the current `post.authors` structure, since `authors` is a collection of objects that also needs to be iterated over in order to find an object with the same `.fname` and `.lname` attributes. I'll do some more research to see if there's a fix, and if not I guess we can merge this one first and change the implementation after a possible refactoring of posts.

MDZHX commented:
It seems that liquid doesn't have an elegant built-in way to check for an object in an array of objects.

I also tried to nest the filters like this:
`{% assign authors_posts = site.posts | where_exp:"post", "{{ post.authors | find_exp:"author","author.fname == page.firstname and author.lname == page.lastname" }} == nil" | sort_by:"date" %}`
and found that liquid doesn't allow that.

So for now, I'm thinking of two ways to implement this without refactoring posts:
* One way is to first manually create an iterable `authors_posts` using nested for loops + if conditions, and then check its size/iterate through it to render the cards. 
* Alternatively, we can use the current boolean flag version.

I wonder what are your thoughts on this? @malsf21 Thanks!

malsf21 reviewed (approved):
Just talked to Zihan, this looks good to merge! We'll look for refactors + fixing more info in another PR.
